File: tui/shared/src/scala/tui/stepBy.scala
package tui

import scala.reflect.ClassTag

object stepBy {
  implicit class StepBySyntax[T](private val ts: Array[T]) extends AnyVal {
    def stepBy(n: Int)(implicit CT: ClassTag[T]): Array[T] = {
      require(n > 0)
      val b = Array.newBuilder[T]
      var i = 0
      while (i < ts.length) {
        b += ts(i)
        i += n
      }
      b.result()
    }
  }
}

File: tui/src/java/tui/internal/StepBy.java
package tui.internal;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.List;

public final class StepBy {
  private StepBy() {}

  @SuppressWarnings("unchecked")
  public static <T> T[] stepBy(T[] ts, int n) {
    if (n <= 0) {
      throw new IllegalArgumentException("n must be positive");
    }
    List<T> out = new ArrayList<>();
    int i = 0;
    while (i < ts.length) {
      out.add(ts[i]);
      i += n;
    }
    T[] result = (T[]) Array.newInstance(ts.getClass().getComponentType(), out.size());
    return out.toArray(result);
  }
}
